File: homework/ba-nguyen/lesson-7/path-sum-ii.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function pathSum(root: TreeNode | null, targetSum: number): number[][] {
  if (!root) return [];

  const nodeValList: number[] = [];
  const pathList: number[][] = [];
  return recursivePathSum(root, targetSum, nodeValList, pathList);
}

function recursivePathSum(
  root: TreeNode | null,
  targetSum: number,
  nodeValList: number[],
  pathList: number[][]
): number[][] {
  if (!root) return pathList;

  nodeValList.push(root.val);

  if (!root.left && !root.right && root.val === targetSum) {
    pathList.push([...nodeValList]);
  }

  recursivePathSum(root.left, targetSum - root.val, nodeValList, pathList);
  recursivePathSum(root.right, targetSum - root.val, nodeValList, pathList);

  nodeValList.pop();

  return pathList;
}

File: homework/Van_Le/homework7.1/PathSumII.cs
public class PathSumII{
    public IList<IList<int>> PathSum(TreeNode root, int targetSum) {
        var result = new List<IList<int>>();
        var currentPath = new Stack<int>();
        DFS(root, result, currentPath, targetSum);
        return result;
    }

    private void DFS(TreeNode node, IList<IList<int>> result, Stack<int> currentPath, int targetSum)
    {
        if (node == null)
            return;
            
        targetSum -= node.val;
        currentPath.Push(node.val);
            
        if (node.left == null && node.right == null)
        {
            if (targetSum == 0)
                result.Add(currentPath.Reverse().ToArray());                
        }
            
        DFS(node.left, result, currentPath, targetSum);
        DFS(node.right, result, currentPath, targetSum);
        currentPath.Pop();            
    }
}

public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}